pints/_abc/_abc_smc.py

```diff
@@ -0,0 +1,198 @@
+#
+# ABC SMC method
+#
+# This file is part of PINTS (https://github.com/pints-team/pints/) which is
+# released under the BSD 3-clause license. See accompanying LICENSE.md for
+# copyright notice and full license details.
+#
+import pints
+import numpy as np
+
+
+class ABCSMC(pints.ABCSampler):
+    r"""
+    Implements the ABC-SMC algorithm as describd in [1].
+
+    In each iteration of the algorithm, the following steps occur:
+
+    .. math::
+        \begin{align}
+        & \theta^* \sim p_{t-1}(\theta) \textrm{, i.e. sample parameters from
+        previous intermediate distribution} \\
+        & \theta^{**} \sim K(\theta|\theta^{*}), \textrm{i.e. perturb }
+        \theta^{*} \textrm{ to     obtain to new point } x \sim
+        p(x|\theta^{**})\textrm{, i.e. sample data from sampling
+        distribution} \\
+        & \textrm{if } s(x) < \textrm{threshold}_(t), \theta^* \textrm{
+        added to list of samples[t]}
+        \end{align}
+
+    After we have obtained nr_samples samples, t is advanced, and weights
+    are calculated for samples[t-1]. At the last error threshold,
+    samples are returned whenever they are accepted. This algorithm is
+    also referred to as ABC Population Monte Carlo (ABC PMC) [2].
+
+    References
+    ----------
+    .. [1] "Toni, Tina, et al. Approximate Bayesian computation scheme
+            for parameter inference and model selection in dynamical systems.
+            Journal of the Royal Society Interface, 6.31: 187-202, 2009.
+            https://doi.org/10.1098/rsif.2008.0172
+
+    .. [2] "Beaumont, Mark A., et al. Adaptive approximate Bayesian
+            computation. Biometrika, 96.4: 983-990, 2009."
+            https://doi.org/10.48550/arXiv.0805.2256
+
+    Parameters
+    ----------
+    nr_samples
+        The number of samples requested for intermediate distributions.
+    error_schedule
+        The schedule of error threshold distance for all distributions.
+    """
+
+    def __init__(self, log_prior, perturbation_kernel=None):
+        # Log prior
+        self._log_prior = log_prior
+
+        # Default value for error threshold schedule
+        self._e_schedule = [1]
+
+        # Default value for current threshold
+        self._threshold = 1
+
+        # Size of intermediate distributions
+        self._nr_samples = 100
+
+        # Set up for first iteration
+        self._samples = [[]]
```

Comment:
We need to think about this. The controller holds the samples too, so we're doubling our memory usage!

Should we update the ask() method so that the user (e.g. the controller) has to provide the samples list?

Reply:
Indeed holding the samples in the `ABCSMC` class as well is a waste. However, after some thinking I believe that changing the ask and tell functions is not the best solution because
- It would change the ask and tell interface for all ABC algorithms (and some algorithms don't need access to the samples) as well as the ABC controller (as it would need to pass on the last generation of samples in the ask and tell calls).
- After finishing a generation of samples, the `ABCSMC` (see the `_advance_time` and `_calculate_weights` functions) also needs access to the current generation of samples, not only the last one.

I think it might be better to simply store in the `ABCSMC` class the current and last generation so that they can be used. In this case there will be some memory duplication, but if we have 10 generations, only the last 2 generations will be stored. Let me know if this seems reasonable.

Reply:
Ok! And your idea of how the schedules should work means that the last 2 generates would be small, e.g. 2000 samples each instead of 10000?

Reply:
I suppose if memory is an issue the controller can be told to stop logging in memory and only log on disk, so then this overhead becomes more bearable

Reply:
Cool! Is there a performance benefit to pre-allocation self._samples? We know it's going to be a matrix of `2 x nr_samples x n_parameters` so could create a numpy array of the correct size and then fill that in, instead of appending to liists?